### PR TITLE
Feature/digest id filter

### PR DESCRIPTION
**Pull request by pati-ni**
Feature for https://github.com/OpenMS/OpenMS/issues/2004

### Brief: 
Works as a feature of TOPP IDFilter. Takes as input a fasta file peptide evidences. It matches peptides evidences with fasta file's proteins by their accession. Then digests the selected protein, and checks whether the peptide evidence is a valid digestion product.

### Testing
It lacks some more substantive Test Files.
The input file is a copy of ./src/tests/topp/THIRDPARTY/MSGFPlusAdapter_1_out.idXML which contains no PeptideEvidences to compare.
@timosachsenberg

### CODE REVIEW

**Comment by timosachsenberg** [comment]
The idXML file does not 100% reflect the class hierarchy (to save space and for historical reasons) .
The PeptideEvidences are contained as part of the PeptideHit.

```
		<PeptideIdentification score_type="SpecEValue" higher_score_better="false" significance_threshold="0" MZ="1063.20983886719" RT="4587.669" spectrum_reference="spectrum=1" >
			<PeptideHit score="3.864406e-26" sequence="IALSRPNVEVVALNDPFITNDYAAYM(Oxidation)FK" charge="3" 
```
**_aa_before="-" aa_after="E" start="1" end="28" protein_refs="PH_1">_**
```
...
			</PeptideHit>
```

**Comment by timosachsenberg** [comment]
would be great if you could add class tests (in addition to the Tool tests)
https://github.com/OpenMS/OpenMS/blob/develop/src/tests/class_tests/openms/source/EnzymaticDigestion_test.cpp
Here you can easier verify if the new isValidProduct works as expected.

**Comment by pati-ni** [comment]
This is the output:
```
$ bin/IDFilter -in $TOPPDATA/IDFilter_12_input.idXML -out /tmp/IDFilter_12_output.idXML -digest:fasta $TOPPDATA/IDFilter_12_input.fasta
Filtering peptides by digested protein (FASTA input)...
Removing unreferenced protein hits...
Before filtering:
1 protein identification(s) with 3 protein hit(s),
3 peptide identification(s) with 3 peptides hit(s).
After filtering:
1 protein identification(s) with 0 protein hit(s),
3 peptide identification(s) with 3 peptides hit(s).
IDFilter took 0.00 s (wall), 0.00 s (CPU), 0.00 s (system), 0.00 s (user).
Accession  is not found in fasta file. Peptide not filtered
```

**Comment by timosachsenberg** [comment]
Thanks. I will be off for today but probably look at it tomorrow.

**Comment by timosachsenberg** [comment]
Can you add another test with missed cleavage 1 and  "Trypsin/P" as enzyme? This should reduce the number of reported hits in the output.

**Comment by pati-ni** [comment]
Ok @timosachsenberg ,
This works, unfortunately I will be able to test it more in the evening.

**Comment by pati-ni** [comment]
@timosachsenberg I added tests and I believe I am done for now. 
However I have one final concern: The  peptide evidence [N_TERMINAL_POSITION](https://github.com/OpenMS/OpenMS/blob/2bcec96a828df8b7d10376bcf644936ecb14e52c/src/openms/source/METADATA/PeptideEvidence.cpp#L42) constant is set to zero. 
I use [hasValidLimits](https://github.com/pati-ni/OpenMS/blob/feature/digest-IDFilter/src/openms/source/METADATA/PeptideEvidence.cpp#L100) in IDFilter checking for this case to skip the peptide evidence checking.
However zero is a perfectly valid parameter for isValidProduct's pep_pos (but not for pep_size). Should I remove [this line](https://github.com/pati-ni/OpenMS/blob/feature/digest-IDFilter/src/openms/source/METADATA/PeptideEvidence.cpp#L105)?

**Comment by timosachsenberg** [comment]
@pati-ni you are right. 0 is a valid position and the line should be removed.
You can get more details on failed tests in http://cdash.openms.de/index.php?project=OpenMS
and if you look at the build from yesterday:
http://cdash.openms.de/viewBuildError.php?buildid=219701

**Comment by hroest** [comment]
@pati-ni there are still some issues, please fix these:

- http://cdash.openms.de/viewBuildError.php?buildid=219777
- http://cdash.openms.de/viewBuildError.php?buildid=219776

**Comment by pati-ni** [comment]
@hroest I'm aware of the issues. Will be fixed today.

File: src/openms/include/OpenMS/CHEMISTRY/EnzymaticDigestion.h [hunk 1]
```diff
@@ -113,7 +113,7 @@ namespace OpenMS
     Size peptideCount(const AASequence& protein);
 
     /// Returns true if peptide at position @p pep_pos with length @p pep_length within protein @p protein was generated by the current model
-    bool isValidProduct(const AASequence& protein, Size pep_pos, Size pep_length, bool methionine_cleavage = false);
+    bool isValidProduct(const AASequence& protein, Size pep_pos, Size pep_length, bool methionine_cleavage = false, bool ignore_missed_cleavages = true) const;
```
Comment (hroest) [inline src/openms/include/OpenMS/CHEMISTRY/EnzymaticDigestion.h]: can you change the `src/pyOpenMS/pxds/EnzymaticDigestion.pxd` signature accordingly to 

```
bool isValidProduct(AASequence protein, Size pep_pos, Size pep_length, bool methionine_cleavage, bool ignore_missed_cleavages) nogil except +
```

thanks

Comment (hroest) [inline src/openms/include/OpenMS/CHEMISTRY/EnzymaticDigestion.h]: I have adopted src/pyOpenMS/pxds/EnzymaticDigestion.pxd  below accordingly

File: src/openms/include/OpenMS/METADATA/PeptideEvidence.h [hunk 2]
```diff
@@ -83,6 +83,8 @@ namespace OpenMS
     /// not equal
     bool operator!=(const PeptideEvidence& rhs) const;
 
+    bool hasValidLimits() const;
```
Comment (hroest) [inline src/openms/include/OpenMS/METADATA/PeptideEvidence.h]: could you add some comments here what this does and what it returns?

File: src/topp/IDFilter.cpp [hunk 3]
```diff
@@ -404,6 +427,50 @@ class TOPPIDFilter :
...
+    String protein_fasta = getStringOption_("digest:fasta").trim();
+    if (!protein_fasta.empty())
+    {
+      LOG_INFO << "Filtering peptides by digested protein (FASTA input)..." << endl;
+      // load protein accessions from FASTA file:
+      vector<FASTAFile::FASTAEntry> fasta;
+      FASTAFile().load(protein_fasta, fasta);
+
+      // Configure Enzymatic digestion
+      // TODO(Nikos) What about missed_cleavages?
```
Comment (hroest) [inline src/topp/IDFilter.cpp]: can you remove or address the TODOs here?

File: src/openms/include/OpenMS/FILTERING/ID/IDFilter.h [hunk 4]
```diff
@@ -282,7 +335,68 @@ namespace OpenMS
...
+                      EnzymaticDigestion& digestion, 
+                      bool ignore_missed_cleavages) : 
+        accession_resolver_(entries), 
+        digestion_(digestion),
+        ignore_missed_cleavages_(ignore_missed_cleavages)
+      {}
+ 
+      bool operator()(const PeptideEvidence& evidence) const
+      {
+        // TODO(Nikos) take into consideration methionine_cleavage parameter (default false)
```
Comment (hroest) [inline src/openms/include/OpenMS/FILTERING/ID/IDFilter.h]: can you address this TODO or remove it?

**Comment by hroest** [comment]
This looks very useful, thanks a lot for this. I have a few comments about the code but to me this looks very good and will soon be ready to be merged into develop

**Comment by pati-ni** [comment]
@hroest Unfortunately I am away from home right now, will be able to make these changes by Friday.

**Comment by pati-ni** [comment]
Requested changes complete. @hroest open to hear any comments about the code. Some diff tests were failing due to [an update in the idXML metadata](http://cdash.openms.de/testDetails.php?test=39530393&build=220445) . Fixed file version and NamespaceSchemaLocation by hand on the TOPP output files and tests passed.

**Comment by hroest** [comment]
:+1:  lets merge

**Pull request merged**